### ISSUE TITLE
Add support for multithreaded hashing

### ISSUE THREAD

**Issue by Quantum-cross** [SmokeMonsterPacks/EverDrive-Packs-Lists-Database#5]
This script would really benefit from this.

**Comment by frederic-mahe** [comment]
In my experience, parsing and building is fast (less than 30 seconds for the N8 pack) if you have a SSD. Multithreaded hashing can speed up things a bit, but my guess is that the main bottleneck is in the speed of your disk, not in the number of CPU cores you have.

I can give multithreading a try nonetheless.

One approach would be to spread the hashing of one file on several CPU cores. I don't think that's possible. Hashing is a serial process and it cannot be split into independent chunks.

The other approach would be to hash one file per CPU core. That's OK for the build script, but it will produce out-of-order results for the parse script: line order would change from one run to another, and git would see that as edits. Not good.

So, I'll try to multithread the build pack, but keep in mind that multithreading is not python's forte. You might be disappointed by the results.

**Comment by Quantum-cross** [comment]
I was building the N64 pack and it was taking quite a while. I remember there were rumblings for packs for disc based consoles (saturn, dreamcast?). I'd imagine that will take quite a while. I'll mess around with it later and see if I am limited by I/O.

**Comment by frederic-mahe** [comment]
You are right to worry about future packs for disc based consoles. Reading a CR-ROM image to get a hash value, or copying a CR-ROM image to several locations will be slow, but in my opinion it is an I/O problem. Faster disks can help, or flash carts with more sophisticated file systems allowing for hard links (like on Unix/Linux) or some sort of file "deduplication" (I don't know if it is possible with FAT32).

One thing I can try in the build pack is to pool reading/writing operations. When copying a ROM to the newly created folders, maybe it is possible to read the ROM only once and to write it at different places simultaneously?

**Comment by FiZiX** [comment]
I believe you're right about the hashing operations being limited by I/O, at least on spinning disk systems. I'm working off of a traditional HDD (spinning disk) and my CPU barely shows any load while building a pack. However, the script still takes over 10 minutes to build from Super EverDrive & SD2SNES SMDB.txt.

**Comment by frederic-mahe** [comment]
I ran some tests.

Hashing a large file stored in RAM, I can reach 99% usage of a CPU core for `sha256sum` (the GNU/Linux command, not the python function). The hashing command then consumes appr. 200 MB/s. My interpretation is that, roughly speaking, as long as your hard drives cannot read and write at least twice that amount of data continuously (400 MB/s), then multithreaded hashing will not really speed up parsing or building.

If you are OK with that, I will mark that issue as a `wontfix` and close it.
